src/asl.ts

```diff
@@ -2129,33 +2319,152 @@ export class ASL {
           ErrorCodes.Cannot_perform_arithmetic_or_bitwise_computations_on_variables_in_Step_Function,
           `Step Function does not support operator ${expr.op}`
         );
-      } else if (
-        expr.op === "instanceof" ||
-        // https://github.com/functionless/functionless/issues/393
-        expr.op === "??=" ||
-        expr.op === "||="
-      ) {
+      } else if (expr.op === "instanceof") {
         throw new SynthError(
           ErrorCodes.Unsupported_Feature,
           `Step Function does not support ${expr.op} operator`
         );
-      }
-      assertNever(expr.op);
-    } else if (isAwaitExpr(expr)) {
-      return this.eval(expr.expr);
-    } else if (isTypeOfExpr(expr)) {
-      return this.evalExpr(expr.expr, (exprOutput) => {
-        if (ASLGraph.isLiteralValue(exprOutput)) {
-          return {
-            value: typeof exprOutput.value,
-            containsJsonPath: false,
-          };
-        }
-
-        const tempHeap = this.newHeapVariable();
+      } else if (expr.op === "&&" || expr.op === "||" || expr.op === "??") {
+        return this.evalExpr(expr.left, (leftOutput) => {
+          const right = this.eval(expr.right);
+
+          if (ASLGraph.isLiteralValue(leftOutput)) {
+            // if both values are literals, return a literal.
+            // Only evaluate right as a literal if it is strictly a literal with no states.
+            if (ASLGraph.isLiteralValue(right)) {
+              return {
+                value:
+                  expr.op === "&&"
+                    ? leftOutput.value && right.value
+                    : expr.op === "||"
+                    ? leftOutput.value || right.value
+                    : leftOutput.value ?? right.value,
+                containsJsonPath: false,
+              };
+            }
 
-        return {
-          startState: "choose",
+            /**
+             * If left is a literal, evaluate the truthiness and return left or right.
+             * &&: when truthy, return right
+             * ||: when falsy, return right
+             * ??: when undefined or null, return right
+             */
+            if (expr.op === "&&") {
+              return !leftOutput.value ? leftOutput : right;
+            } else if (expr.op === "||") {
+              return !!leftOutput.value ? leftOutput : right;
```

Comment:
I find it a bit challengin to grok `!!`, any chance you can change this into a more explicit comparison?

Reply:
`!!` and `Boolean(...)` are the most succinct ways to handle truthy in JS. I wouldn't want to re-implement truthy logic (and we actually want truthy here).

Reply:
What does !!` do?

Reply:
`!` not truthy, `!!` is truthy.

> It is possible to use a couple of NOT operators in series to explicitly force the conversion of any value to the corresponding [boolean primitive](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#boolean_type). The conversion is based on the "[truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy)ness" or "[falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy)ness" of the value (see truthy and falsy).

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_NOT#double_not_!!

> Do not use a Boolean object to convert a non-boolean value to a boolean value. To perform this task, instead, use Boolean as a function, or a [double NOT operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_NOT#double_not_!!):

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean

Reply:
That is unnecessary on this case

Reply:
Yeah, I think you are right, the ternary should also enforce truthiness. I'll change it in the next PR.